File: 백준/Gold/2252. 줄 세우기/줄 세우기.py
n, m = map(int, input().split())

adj = [[] for _ in range(n + 1)]
ind = [0] * (n + 1) # 들어오는 간선의 개수 in-degree

for _ in range(m):
    x, y = map(int, input().split())
    adj[x].append(y)
    ind[y] += 1

candidates = []
for i in range(1, n + 1):
    if ind[i] == 0:
        candidates.append(i)

result = []

while candidates:
    cur = candidates.pop()
    result.append(cur)
    
    for nxt in adj[cur]:
        ind[nxt] -= 1
        if ind[nxt] == 0:
            candidates.append(nxt)

print(*result)

File: 백준/Gold/2252. 줄 세우기/줄 세우기.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int m = sc.nextInt();

        ArrayList<ArrayList<Integer>> adj = new ArrayList<>();
        for (int i = 0; i <= n; i++) {
            adj.add(new ArrayList<>());
        }
        int[] ind = new int[n + 1];

        for (int i = 0; i < m; i++) {
            int x = sc.nextInt();
            int y = sc.nextInt();
            adj.get(x).add(y);
            ind[y]++;
        }

        ArrayList<Integer> candidates = new ArrayList<>();
        for (int i = 1; i <= n; i++) {
            if (ind[i] == 0) {
                candidates.add(i);
            }
        }

        ArrayList<Integer> result = new ArrayList<>();
        while (!candidates.isEmpty()) {
            int cur = candidates.remove(candidates.size() - 1);
            result.add(cur);

            for (int nxt : adj.get(cur)) {
                ind[nxt]--;
                if (ind[nxt] == 0) {
                    candidates.add(nxt);
                }
            }
        }

        for (int num : result) {
            System.out.print(num + " ");
        }
    }
}